docs/src/tutorials/deim_FitzHugh-Nagumo.md

```diff
@@ -0,0 +1,192 @@
+vec_v, sol_v = splitview(split_vec_sizes...)
+vec_w, sol_w = splitview(split_vec_sizes...)
+for (i, (xᵢ, vᵢ, wᵢ)) in enumerate(zip(grid_x, grid_v, grid_w))
+    iₓ₂ = 2i
+    iₓ₂₋₁ = iₓ₂ - 1
+    sol_x[iₓ₂₋₁] .= xᵢ
+    sol_x[iₓ₂] .= NaN # NaN is a path separator for plotting unconnected data
+    sol_v[iₓ₂₋₁] .= sol[vᵢ]
+    sol_v[iₓ₂] .= NaN
+    sol_w[iₓ₂₋₁] .= sol[wᵢ]
+    sol_w[iₓ₂] .= NaN
+end
+full_plt = plot(xlabel = L"v(x,t)", ylabel = L"x", zlabel = L"w(x,t)", xlims = (-0.5, 2.0),
+           ylims = (0.0, L), zlims = (0.0, 0.25), legend = false, xflip = true,
+           camera = (50, 30), titlefont = 10,
+           title = "Phase−Space diagram of full $(nameof(pde_sys)) system")
+for i in 1:2:(2n_x)
+    plot!(full_plt, sol_v[i], sol_x[i], sol_w[i])
+end
+plot!(full_plt)
+```
+
+Then, we apply POD-DEIM of dimension, say, 5 and solve the reduced system.
+
+```@example deim_FitzHugh_Nagumo
+using ModelOrderReduction
+snapshot_simpsys = Array(sol)
+pod_dim = deim_dim = 5
+pod_reducer = POD(snapshot_simpsys, pod_dim)
+reduce!(pod_reducer, TSVD())
+pod_basis = pod_reducer.rbasis
+deim_sys = deim(simp_sys, pod_basis)
+deim_prob = ODEProblem(deim_sys, nothing, tspan)
+deim_sol = solve(deim_prob)
+n_t_deim = length(deim_sol.t)
+nothing # hide
+```
+
+And plot the result from the POD-DEIM reduced system.
+
+```@example deim_FitzHugh_Nagumo
+split_vec_sizes_deim = collect(isodd(i) ? (n_t_deim,) : (1,) for i in 1:(2n_x))
+vec_deim_x, sol_deim_x = splitview(split_vec_sizes_deim...)
+vec_deim_v, sol_deim_v = splitview(split_vec_sizes_deim...)
+vec_deim_w, sol_deim_w = splitview(split_vec_sizes_deim...)
+for (i, (xᵢ, vᵢ, wᵢ)) in enumerate(zip(grid_x, grid_v, grid_w))
+    iₓ₂ = 2i
+    iₓ₂₋₁ = iₓ₂ - 1
+    sol_deim_x[iₓ₂₋₁] .= xᵢ
+    sol_deim_x[iₓ₂] .= NaN
+    sol_deim_v[iₓ₂₋₁] .= deim_sol[vᵢ]
+    sol_deim_v[iₓ₂] .= NaN
+    sol_deim_w[iₓ₂₋₁] .= deim_sol[wᵢ]
+    sol_deim_w[iₓ₂] .= NaN
+end
+deim_plt = plot(xlabel = L"v(x,t)", ylabel = L"x", zlabel = L"w(x,t)", xlims = (-0.5, 2.0),
+                ylims = (0.0, L), zlims = (0.0, 0.25), legend = false, xflip = true,
+                camera = (50, 30), titlefont = 10,
+                title = "Phase−Space diagram of reduced $(nameof(pde_sys)) system")
+for i in 1:2:(2n_x)
+    plot!(deim_plt, sol_deim_v[i], sol_deim_x[i], sol_deim_w[i])
+end
+plot!(deim_plt)
+```
+
+Finally, we put the two solutions in one figure.
+```@example deim_FitzHugh_Nagumo
+plt_2 = plot(xlabel = L"v(x,t)", ylabel = L"x", zlabel = L"w(x,t)", xlims = (-0.5, 2.0),
+             ylims = (0.0, L), zlims = (0.0, 0.25), xflip = true, camera = (50, 30), 
+             titlefont = 10, title = "Comparison of full and reduced systems")
+plot!(plt_2, vec_v, vec_x, vec_w, label = "Full$(length(ode_sys.eqs))")
+plot!(plt_2, vec_deim_v, vec_deim_x, vec_deim_w, label = "POD$(pod_dim)/DEIM$(deim_dim)")
+```
+
+As we can see, the reduced-order system captures the limit cycle of the original full-order 
+system very well.
```

Comment:
Yes, I should add that. The DEIM paper sets the dimension to 1024 in the full-order model.

However, at the moment, I cannot easily do benchmark.

```julia
using ModelingToolkit
@variables t x(t) y(t)
D = Differential(t)
A = [3; 4]
s = x + y
eqs = D.([x, y]) .~ A * s
@named sys = ODESystem(eqs)
u0 = Dict(x => 5.0, y => 6.0)
tspan = (0.0, 1.0)
prob = ODEProblem(sys, u0, tspan)
prob.f.f.f_iip.body
```

yields

```julia
quote
    begin
        begin
            #= /home/bowen/.julia/packages/SymbolicUtils/qulQp/src/code.jl:398 =# @inbounds begin
                    ˍ₋out[1] = (+)((*)(3, (getindex)(ˍ₋arg1, 1)), (*)(3, (getindex)(ˍ₋arg1, 2)))
                    ˍ₋out[2] = (+)((*)(4, (getindex)(ˍ₋arg1, 1)), (*)(4, (getindex)(ˍ₋arg1, 2)))
                    nothing
                end
        end
    end
end
```

The same addition is evaluated twice in the above function generated at rumtime.

(JuliaSymbolics/SymbolicUtils.jl#200) The argument `cse`, common subexpression elimination, in `Symbolics.build_function` is indeed a remedy but it's slow and we cannot rely on that. 

We might need to wait for some ongoing changes in ModelingToolkit and Symbolics.

Reply:
I am wondering if a neat workaround could be the following (that probably goes beyond your GSoC project but I think it is interesting to discuss): 

We use symbolics/MTK to generate a function only for evaluation of the relevant entries of the nonlinear function (at the DEIM indices). And then we simply use standard codegen to generate a non-allocating RHS for the reduced model since the structure of that would always be the same: Something along the lines of 
``` 
function pod_deim_rhs(du, u, (cached_state, cached_nonlin, V_deim, A, g, P_deim), t)
mul!(cached_state, V_deim, u) # V_deim refers to the matrix mapping the reduced state to the full order model states that are needed to evaluate the deim interpolation
deim_nodes!(cached_nonlin, cached_state) # deim_nodes! would be generated via MTK/Symbolics and evaluate the nonlinear part at the deim nodes. 
du .= A * u + g + P_deim*cached_nonlin # where A * u refers to the reduced linear part, g to the reduced constant part and P_deim to the deim projector. 
end
```
Of course there are still quite a few questions about this. Most prominently, I wonder if there is a slicker way than passing caches, and constant coefficient matrices as parameters?  

Have you already thought along those lines and is there anything that speaks against it? Other than worse composition with MTK etc?

Reply:
I did benchmark for a full model of dimension 512 with POD and DEIM dimensions in `[1, 3, 5, 10]`.
https://github.com/bowenszhu/ModelOrderReduction.jl/blob/DEIM-benchmark/benchmark.jl

The results are 

![average relative error](https://raw.githubusercontent.com/bowenszhu/ModelOrderReduction.jl/DEIM-benchmark/error.png)

![time construct ODEProblem](https://raw.githubusercontent.com/bowenszhu/ModelOrderReduction.jl/DEIM-benchmark/prob_time.png)

![memory construct ODEProblem](https://raw.githubusercontent.com/bowenszhu/ModelOrderReduction.jl/DEIM-benchmark/prob_mem.png)

![time solve ODEProblem](https://raw.githubusercontent.com/bowenszhu/ModelOrderReduction.jl/DEIM-benchmark/sol_time.png)

![memory solve ODEProblem](https://raw.githubusercontent.com/bowenszhu/ModelOrderReduction.jl/DEIM-benchmark/sol_mem.png)

Reply:
that benchmark looks really good. I am a bit surprised that the error is so large. The error is clearly dominated by the mismatch of ROM and FOM at early times. What time horizon did you choose to evaluate the error? What do we see if we look at long times where the system has effectively converged to the limit cycle?

Reply:
1. Previously I encountered both of the following 2 errors when doing benchmarks
```ERROR: LoadError: mul_float: types of a and b must match``` 
SciML/ModelingToolkit.jl#1515
```ERROR: UndefVarError: mul_float not defined``` 
SciML/RuntimeGeneratedFunctions.jl#13
I thought the `deim` function I implemented was not able to handle large models, but it is actually capable. The errors only happen when one defines too many **large** functions at runtime in the same CPU process. Therefore, I modified my benchmark code such that each trial of POD dimension and DEIM interpolation dimension runs in separate processes.

1. Besides dependent variables, the variable shown as an axis in the plots [here](https://github.com/SciML/ModelOrderReduction.jl/pull/19#discussion_r964060923) is space `x` instead of time `t`. The big differences are at the region where `x` is close to 0, the left boundary. I made plots for dimensions up to 40 in [this branch](https://github.com/bowenszhu/ModelOrderReduction.jl/tree/DEIM-plot). The error at the region close to the left boundary is almost eliminated as we can see in [this plot](https://github.com/bowenszhu/ModelOrderReduction.jl/blob/DEIM-plot/POD40_DEIM40.png) where the POD dimension and the DEIM dimension are both 40.

1. After some trial and error, I found that the `ODEProblem`s were not solved to a desired precision. Then, I updated the corresponding solver options for `DifferentialEquations.solve`. The [new plot](https://github.com/bowenszhu/ModelOrderReduction.jl/commit/1784164e89b27a201731d3b82e43f4535b8f0192#commitcomment-84237863) looks better as the error decreases when the dimensions become larger.

Reply:
Although I must admit, I am still a bit confused by your second point: I understand that the independent variable shown explicitly in the plot is `x`. But isn't time `t` implicit? As in each trace in the plot shows the trajectory of the system state for a fixed `x`, i.e., $\lbrace(v(x,t), w(x,t)) : t \in [0,T] \rbrace$? And unless I am really not reading this right, the plots [here](https://github.com/SciML/ModelOrderReduction.jl/pull/19#discussion_r964060923) look like that the ROM is converging to the correct limit cycle close to `x=0` as `t` increases? Can you tell me what I am parsing wrong here? Maybe the plot is too crowded and I just can't identify the right traces?

Reply:
OK I see. I think the issue here is that the error metric you are using is a bit misleading. I think it should be normalized by the length of the vector $\mathbf{y}(t)$ since after all the error metric for the discretized system should approximate the intuitive error metric for the continuous solution: 
$$\frac{\| \mathbf{y}(t) - \mathbf{y}^{deim}(t)\|}{\text{length}(\mathbf{y})} \approx \sqrt{\int_{0}^1 |y(x,t) - y^{deim}(x,t) |^2  \ dx}$$

Correct me if I am wrong, but after skimming through your benchmark code it seems like you are not doing that currently and that would bring the error down by 10^3 which is a lot more in line with what one would expect based on the plots. 

Anyway, this looks all very nice and we are at a point where we are discussing minor minor things here. I think we should move forward.